// thor-devkit/dist/lib/abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ethABI = require('web3-eth-abi');
/** encode/decode parameters of contract function call, event log, according to ABI JSON */
var abi;
(function (abi) {
    /** for contract function */
    class Function {
        /**
         * create a function object
         * @param definition abi definition of the function
         */
        constructor(definition) {
            this.definition = definition;
        }
        /** compute function signature */
        get signature() {
            return ethABI.encodeFunctionSignature(this.definition);
        }
        /**
         * encode input parameters into call data
         * @param args arguments for the function
         */
        encode(...args) {
            return ethABI.encodeFunctionCall(this.definition, args);
        }
        /**
         * decode output data
         * @param outputData output data to decode
         */
        decode(outputData) {
            return ethABI.decodeParameters(this.definition.outputs || [], outputData);
        }
    }
    abi.Function = Function;
    /** for contract event */
    class Event {
        /** for contract event */
        constructor(definition) {
            this.definition = definition;
        }
        /** compute event signature */
        get signature() {
            return ethABI.encodeEventSignature(this.definition);
        }
        /**
         * encode an object of indexed keys into topics.
         * @param indexed an object contains indexed keys
         */
        encode(indexed) {
            const topics = [];
            if (!this.definition.anonymous) {
                topics.push(this.signature);
            }
            for (const input of this.definition.inputs) {
                if (!input.indexed) {
                    continue;
                }
                const value = indexed[input.name];
                if (value) {
                    // TODO: special case for dynamic types
                    topics.push(ethABI.encodeParameter(input.type, value));
                }
                else {
                    topics.push(null);
                }
            }
            return topics;
        }
        /**
         * decode event log
         * @param data data in event output
         * @param topics topics in event
         */
        decode(data, topics) {
            return ethABI.decodeLog(this.definition.inputs, data, this.definition.anonymous ? topics : topics.slice(1));
        }
    }
    abi.Event = Event;
})(abi = exports.abi || (exports.abi = {}));
