// thor-devkit/dist/lib/crypto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("crypto");
const keccak = require('keccak');
const blake = require('blakejs');
/**
 * computes blake2b 256bit hash of given data
 * @param data one or more Buffer | string
 */
function blake2b256(...data) {
    const ctx = blake.blake2bInit(32, null);
    data.forEach(d => {
        if (Buffer.isBuffer(d)) {
            blake.blake2bUpdate(ctx, d);
        }
        else {
            blake.blake2bUpdate(ctx, Buffer.from(d, 'utf8'));
        }
    });
    return Buffer.from(blake.blake2bFinal(ctx));
}
exports.blake2b256 = blake2b256;
/**
 * computes keccak256 hash of given data
 * @param data one or more Buffer | string
 */
function keccak256(...data) {
    const h = keccak('keccak256');
    data.forEach(d => {
        if (Buffer.isBuffer(d)) {
            h.update(d);
        }
        else {
            h.update(Buffer.from(d, 'utf8'));
        }
    });
    return h.digest();
}
exports.keccak256 = keccak256;
/**
 * derive Address from public key
 * @param pubKey the public key
 */
function publicKeyToAddress(pubKey) {
    return keccak256(pubKey.slice(1)).slice(12);
}
exports.publicKeyToAddress = publicKeyToAddress;
const secp256k1Funs = require('secp256k1');
/** secp256k1 methods set */
var secp256k1;
(function (secp256k1) {
    /** generate private key  */
    function generatePrivateKey() {
        for (;;) {
            const privKey = crypto_1.randomBytes(32);
            if (secp256k1Funs.privateKeyVerify(privKey)) {
                return privKey;
            }
        }
    }
    secp256k1.generatePrivateKey = generatePrivateKey;
    /**
     * derive public key(uncompressed) from private key
     * @param privKey the private key
     */
    function derivePublicKey(privKey) {
        return secp256k1Funs.publicKeyCreate(privKey, false /* uncompressed */);
    }
    secp256k1.derivePublicKey = derivePublicKey;
    /**
     * sign a message using elliptic curve algorithm on the curve secp256k1
     * @param msgHash hash of message
     * @param privKey serialized private key
     */
    function sign(msgHash, privKey) {
        const sig = secp256k1Funs.sign(msgHash, privKey);
        const packed = Buffer.alloc(65);
        sig.signature.copy(packed);
        packed[64] = sig.recovery;
        return packed;
    }
    secp256k1.sign = sign;
    /**
     * recovery signature to public key
     * @param msgHash hash of message
     * @param sig signature
     */
    function recover(msgHash, sig) {
        if (sig.length !== 65) {
            throw new Error('invalid signature');
        }
        const recovery = sig[64];
        if (recovery !== 0 && recovery !== 1) {
            throw new Error('invalid signature recovery');
        }
        return secp256k1Funs.recover(msgHash, sig.slice(0, 64), recovery, false);
    }
    secp256k1.recover = recover;
})(secp256k1 = exports.secp256k1 || (exports.secp256k1 = {}));
const Keythereum = require('keythereum');
var Keystore;
(function (Keystore) {
    /**
     * encrypt private key to keystore with given password
     * @param privateKey the private key to be encrypted
     * @param password password to encrypt the private key
     */
    function encrypt(privateKey, password) {
        return new Promise((resolve, reject) => {
            if (!secp256k1Funs.privateKeyVerify(privateKey)) {
                return reject(new Error('invalid private key'));
            }
            Keythereum.dump(password, privateKey, crypto_1.randomBytes(32), crypto_1.randomBytes(16), {
                cipher: 'aes-128-ctr',
                kdf: 'scrypt',
                kdfparams: {
                    dklen: 32,
                    memory: 280000000,
                    n: 262144,
                    p: 1,
                    r: 8,
                },
            }, resolve);
        });
    }
    Keystore.encrypt = encrypt;
    /**
     * decrypt private key from keystore
     * an error thrown if not well formed
     * @param ks the keystore
     * @param password password to decrypt keystore
     */
    function decrypt(ks, password) {
        return new Promise((resolve, reject) => {
            Keythereum.recover(password, validate(normalize(ks)), (r) => {
                if (!Buffer.isBuffer(r)) {
                    return reject(r);
                }
                resolve(r);
            });
        });
    }
    Keystore.decrypt = decrypt;
    /**
     * roughly check whether keystore is well formed
     * @param ks the keystore
     */
    function wellFormed(ks) {
        try {
            validate(normalize(ks));
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    Keystore.wellFormed = wellFormed;
    /** normalize keystore. e.g. lower case keys */
    function normalize(obj) {
        const lowerKey = (o) => {
            return Object.keys(o).reduce((converted, k) => {
                let v = o[k];
                if (typeof v === 'object') {
                    v = lowerKey(v);
                }
                converted[k.toLowerCase()] = v;
                return converted;
            }, {});
        };
        return lowerKey(obj);
    }
    function validate(ks) {
        if (ks.version !== 1 && ks.version !== 3) {
            throw new Error('unsupported version');
        }
        if (!/^[0-9a-f]{40}$/i.test(ks.address)) {
            throw new Error('invalid address');
        }
        if (!/^[0-9a-z]{8}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{12}$/i.test(ks.id)) {
            throw new Error('invalid id');
        }
        if (typeof ks.crypto !== 'object') {
            throw new Error('invalid crypto');
        }
        return ks;
    }
})(Keystore = exports.Keystore || (exports.Keystore = {}));
const BIP39 = require("bip39");
const HDKey = require("hdkey");
var mnemonic;
(function (mnemonic) {
    // see https://github.com/satoshilabs/slips/blob/master/slip-0044.md
    const VET_DERIVATION_PATH = `m/44'/818'/0'/0`;
    /** generate BIP39 mnemonic words */
    function generate() {
        return BIP39.generateMnemonic(128, crypto_1.randomBytes).split(' ');
    }
    mnemonic.generate = generate;
    /**
     * check if the given mnemonic words have valid checksum
     * @param words mnemonic words
     */
    function validate(words) {
        return BIP39.validateMnemonic(words.join(' '));
    }
    mnemonic.validate = validate;
    /**
     * derive private key from mnemonic words according to BIP32.
     * the derivation path is defined at https://github.com/satoshilabs/slips/blob/master/slip-0044.md
     */
    function derivePrivateKey(words) {
        const seed = BIP39.mnemonicToSeed(words.join(' '));
        const hdKey = HDKey.fromMasterSeed(seed);
        return hdKey.derive(VET_DERIVATION_PATH + '/0').privateKey;
    }
    mnemonic.derivePrivateKey = derivePrivateKey;
})(mnemonic = exports.mnemonic || (exports.mnemonic = {}));
