// thor-devkit/dist/lib/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("./crypto");
const rlp_1 = require("./rlp");
/** Transaction class defines VeChainThor's multi-clause transaction */
class Transaction {
    /** decode from Buffer to transaction */
    static decode(raw) {
        try {
            const body = unsignedTxRLP.decode(raw);
            return new Transaction(body);
        }
        catch (_a) {
            const body = txRLP.decode(raw);
            const sig = body.signature;
            delete body.signature;
            const tx = new Transaction(body);
            tx.signature = Buffer.from(sig.slice(2), 'hex');
            return tx;
        }
    }
    /**
     * construct a transaction object with given body
     * @param body body of tx
     */
    constructor(body) {
        this.body = Object.assign({}, body, { reserved: body.reserved || [] });
    }
    /**
     * returns transaction ID
     * null returned if something wrong (e.g. invalid signature)
     */
    get id() {
        if (!this.signature) {
            return null;
        }
        try {
            const signingHash = crypto_1.blake2b256(unsignedTxRLP.encode(this.body));
            const pubKey = crypto_1.secp256k1.recover(signingHash, this.signature);
            const signer = crypto_1.publicKeyToAddress(pubKey);
            return '0x' + crypto_1.blake2b256(signingHash, signer).toString('hex');
        }
        catch (_a) {
            return null;
        }
    }
    /** returns signer. null returned if no signature or not incorrectly signed */
    get signer() {
        if (!this.signature) {
            return null;
        }
        try {
            const signingHash = crypto_1.blake2b256(unsignedTxRLP.encode(this.body));
            const pubKey = crypto_1.secp256k1.recover(signingHash, this.signature);
            return '0x' + crypto_1.publicKeyToAddress(pubKey).toString('hex');
        }
        catch (_a) {
            return null;
        }
    }
    /** returns intrinsic gas it takes */
    get intrinsicGas() {
        return Transaction.intrinsicGas(this.body.clauses);
    }
    /** encode into Buffer */
    encode() {
        if (this.signature) {
            const sigHex = '0x' + this.signature.toString('hex');
            return txRLP.encode(Object.assign({}, this.body, { signature: sigHex }));
        }
        return unsignedTxRLP.encode(this.body);
    }
}
exports.Transaction = Transaction;
(function (Transaction) {
    /**
     * calculates intrinsic gas that a tx costs with the given clauses.
     * @param clauses
     */
    function intrinsicGas(clauses) {
        const txGas = 5000;
        const clauseGas = 16000;
        const clauseGasContractCreation = 48000;
        if (clauses.length === 0) {
            return txGas + clauseGas;
        }
        return clauses.reduce((sum, c) => {
            if (c.to) {
                sum += clauseGas;
            }
            else {
                sum += clauseGasContractCreation;
            }
            sum += dataGas(c.data);
            return sum;
        }, txGas);
    }
    Transaction.intrinsicGas = intrinsicGas;
    function dataGas(data) {
        const zgas = 4;
        const nzgas = 68;
        let sum = 0;
        for (let i = 2; i < data.length; i += 2) {
            if (data.substr(i, 2) === '00') {
                sum += zgas;
            }
            else {
                sum += nzgas;
            }
        }
        return sum;
    }
})(Transaction = exports.Transaction || (exports.Transaction = {}));
const unsignedTxRLP = new rlp_1.RLP({
    name: 'tx',
    kind: [
        { name: 'chainTag', kind: new rlp_1.RLP.NumericKind(1) },
        { name: 'blockRef', kind: new rlp_1.RLP.TrimmedBlobKind(8) },
        { name: 'expiration', kind: new rlp_1.RLP.NumericKind(4) },
        {
            name: 'clauses', kind: {
                item: [
                    { name: 'to', kind: new rlp_1.RLP.NullableBlobKind(20) },
                    { name: 'value', kind: new rlp_1.RLP.NumericKind(32) },
                    { name: 'data', kind: new rlp_1.RLP.VariableBlobKind() },
                ],
            },
        },
        { name: 'gasPriceCoef', kind: new rlp_1.RLP.NumericKind(1) },
        { name: 'gas', kind: new rlp_1.RLP.NumericKind(8) },
        { name: 'dependsOn', kind: new rlp_1.RLP.NullableBlobKind(32) },
        { name: 'nonce', kind: new rlp_1.RLP.NumericKind(8) },
        { name: 'reserved', kind: new rlp_1.RLP.RawKind() },
    ],
});
const txRLP = new rlp_1.RLP({
    name: 'tx',
    kind: [...unsignedTxRLP.profile.kind, { name: 'signature', kind: new rlp_1.RLP.VariableBlobKind() }],
});
