// thorify/dist/extend/accounts.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3Utils = require('web3-utils');
const debug = require('debug')('thor:injector');
const EthLib = require('eth-lib/lib');
const thor_devkit_1 = require("thor-devkit");
const utils = require("../utils");
const extendAccounts = function (web3) {
    // signTransaction supports both callback and promise style
    web3.eth.accounts.signTransaction = function signTransaction(ethTx, privateKey, callback) {
        debug('tx to sign: %O', ethTx);
        const sign = function (tx) {
            return __awaiter(this, void 0, void 0, function* () {
                if (!tx.chainTag) {
                    const chainTag = yield web3.eth.getChainTag();
                    if (chainTag) {
                        tx.chainTag = chainTag;
                    }
                    else {
                        throw new Error('error getting chainTag');
                    }
                }
                if (!tx.blockRef) {
                    const blockRef = yield web3.eth.getBlockRef();
                    if (blockRef) {
                        tx.blockRef = blockRef;
                    }
                    else {
                        throw new Error('error getting blockRef');
                    }
                }
                if (tx.data && utils.isHex(tx.data)) {
                    tx.data = utils.toPrefixedHex(tx.data);
                }
                else if (tx.data) {
                    throw new Error('Data must be valid hex');
                }
                else {
                    tx.data = '0x';
                }
                if (!tx.gas) {
                    const gas = yield web3.eth.estimateGas({
                        from: EthLib.account.fromPrivate(utils.toPrefixedHex(privateKey)).address,
                        to: tx.to ? tx.to : '',
                        value: tx.value ? tx.value : 0,
                        data: tx.data,
                    });
                    if (gas) {
                        tx.gas = gas;
                    }
                    else {
                        throw new Error('error getting gas');
                    }
                }
                if (!tx.nonce) {
                    tx.nonce = utils.newNonce();
                }
                const clause = {
                    value: tx.value || 0,
                    to: tx.to || null,
                    data: tx.data,
                };
                const body = {
                    chainTag: utils.validNumberOrDefault(tx.chainTag, 0),
                    blockRef: tx.blockRef,
                    gas: tx.gas,
                    expiration: utils.validNumberOrDefault(tx.expiration, utils.params.defaultExpiration),
                    gasPriceCoef: utils.validNumberOrDefault(tx.gasPriceCoef, utils.params.defaultGasPriceCoef),
                    dependsOn: !tx.dependsOn ? null : tx.dependsOn,
                    nonce: typeof tx.nonce === 'string' ? utils.toPrefixedHex(tx.nonce) : tx.nonce,
                    clauses: [clause],
                };
                debug('body: %O', body);
                const ThorTx = new thor_devkit_1.Transaction(body);
                const priKey = Buffer.from(utils.sanitizeHex(privateKey), 'hex');
                const signingHash = thor_devkit_1.cry.blake2b256(ThorTx.encode());
                ThorTx.signature = thor_devkit_1.cry.secp256k1.sign(signingHash, priKey);
                const result = {
                    rawTransaction: utils.toPrefixedHex(ThorTx.encode().toString('hex')),
                    messageHash: signingHash,
                };
                return result;
            });
        };
        // for supporting both callback and promise
        if (callback instanceof Function) {
            sign(ethTx).then((ret) => {
                return callback(null, ret);
            }).catch((e) => {
                return callback(e);
            });
        }
        else {
            return sign(ethTx);
        }
    };
    web3.eth.accounts.recoverTransaction = function recoverTransaction(encodedRawTx) {
        const values = EthLib.RLP.decode(encodedRawTx);
        const signingDataHex = EthLib.RLP.encode(values.slice(0, 9));
        const singingHashBuffer = thor_devkit_1.cry.blake2b256(Buffer.from(utils.sanitizeHex(signingDataHex), 'hex'));
        const signature = values[9];
        const signatureBuffer = Buffer.from(utils.sanitizeHex(signature), 'hex');
        const pubKey = thor_devkit_1.cry.secp256k1.recover(singingHashBuffer, signatureBuffer);
        const address = thor_devkit_1.cry.publicKeyToAddress(pubKey);
        return utils.toPrefixedHex(address.toString('hex'));
    };
    web3.eth.accounts.hashMessage = function hashMessage(data) {
        const message = web3Utils.isHexStrict(data) ? web3Utils.hexToBytes(data) : data;
        const messageBuffer = Buffer.from(message);
        return utils.toPrefixedHex(thor_devkit_1.cry.blake2b256(messageBuffer).toString('hex'));
    };
    web3.eth.accounts.sign = function sign(data, privateKey) {
        const hash = this.hashMessage(data);
        const hashBuffer = Buffer.from(utils.sanitizeHex(hash), 'hex');
        const privateKeyBuffer = Buffer.from(utils.sanitizeHex(privateKey), 'hex');
        const signature = thor_devkit_1.cry.secp256k1.sign(hashBuffer, privateKeyBuffer).toString('hex');
        return {
            message: data,
            messageHash: utils.toPrefixedHex(hash),
            signature: utils.toPrefixedHex(signature),
        };
    };
    web3.eth.accounts.recover = function recover(message, signature, preFixed) {
        const args = [].slice.apply(arguments);
        if (utils.isObject(message)) {
            return this.recover(message.messageHash, message.signature, true);
        }
        if (!preFixed) {
            message = this.hashMessage(message);
        }
        const hexBuffer = Buffer.from(utils.sanitizeHex(message), 'hex');
        const signatureBuffer = Buffer.from(utils.sanitizeHex(signature), 'hex');
        const pubKey = thor_devkit_1.cry.secp256k1.recover(hexBuffer, signatureBuffer);
        const address = thor_devkit_1.cry.publicKeyToAddress(pubKey);
        return utils.toPrefixedHex(address.toString('hex'));
    };
};
exports.extendAccounts = extendAccounts;
